# Introduction to Pandas Lab

Complete the following set of exercises to solidify your knowledge of Pandas fundamentals.

### 1. Import Numpy and Pandas and alias them to `np` and `pd` respectively.

In [1]:
import pandas as pd
import numpy as np

### 2. Create a Pandas Series containing the elements of the list below.

In [4]:
lst = [5.7, 75.2, 74.4, 84.0, 66.5, 66.3, 55.8, 75.7, 29.1, 43.7]
lst = pd.Series(lst) 

### 3. Use indexing to return the third value in the Series above.

*Hint: Remember that indexing begins at 0.*

In [5]:
lst[2]

np.float64(74.4)

### 4. Create a Pandas DataFrame from the list of lists below. Each sublist should be represented as a row.

In [6]:
b = [[53.1, 95.0, 67.5, 35.0, 78.4],
     [61.3, 40.8, 30.8, 37.8, 87.6],
     [20.6, 73.2, 44.2, 14.6, 91.8],
     [57.4, 0.1, 96.1, 4.2, 69.5],
     [83.6, 20.5, 85.4, 22.8, 35.9],
     [49.0, 69.0, 0.1, 31.8, 89.1],
     [23.3, 40.7, 95.0, 83.8, 26.9],
     [27.6, 26.4, 53.8, 88.8, 68.5],
     [96.6, 96.4, 53.4, 72.4, 50.1],
     [73.7, 39.0, 43.2, 81.6, 34.7]]

In [ ]:
df = pd.DataFrame(b)


### 5. Rename the data frame columns based on the names in the list below.

```py
list_names = ["Score_1", "Score_2", "Score_3", "Score_4", "Score_5"]
```

In [9]:
b = [[53.1, 95.0, 67.5, 35.0, 78.4],
     [61.3, 40.8, 30.8, 37.8, 87.6],
     [20.6, 73.2, 44.2, 14.6, 91.8],
     [57.4, 0.1, 96.1, 4.2, 69.5],
     [83.6, 20.5, 85.4, 22.8, 35.9],
     [49.0, 69.0, 0.1, 31.8, 89.1],
     [23.3, 40.7, 95.0, 83.8, 26.9],
     [27.6, 26.4, 53.8, 88.8, 68.5],
     [96.6, 96.4, 53.4, 72.4, 50.1],
     [73.7, 39.0, 43.2, 81.6, 34.7]]

In [12]:
list_names = ["Score_1", "Score_2", "Score_3", "Score_4", "Score_5"]
df.columns = list_names

### 6. Create a subset of this data frame that contains only the Score 1, 3, and 5 columns.

In [19]:
df[['Score_1', 'Score_3', 'Score_5']]

,Score_1,Score_3,Score_5
0,53.1,67.5,78.4
1,61.3,30.8,87.6
2,20.6,44.2,91.8
3,57.4,96.1,69.5
4,83.6,85.4,35.9
5,49.0,0.1,89.1
6,23.3,95.0,26.9
7,27.6,53.8,68.5
8,96.6,53.4,50.1
9,73.7,43.2,34.7


### 7. From the original data frame, calculate the average Score_3 value.

In [21]:
df.mean()['Score_3']

np.float64(56.95000000000001)

### 8. From the original data frame, calculate the maximum Score_4 value.

In [22]:
df.max()['Score_4']

np.float64(88.8)

### 9. From the original data frame, calculate the median Score 2 value.

In [23]:
df.median()['Score_2']

np.float64(40.75)

### 10. Create a Pandas DataFrame from the dictionary of product orders below.

In [24]:
orders = {'Description': ['LUNCH BAG APPLE DESIGN',
  'SET OF 60 VINTAGE LEAF CAKE CASES ',
  'RIBBON REEL STRIPES DESIGN ',
  'WORLD WAR 2 GLIDERS ASSTD DESIGNS',
  'PLAYING CARDS JUBILEE UNION JACK',
  'POPCORN HOLDER',
  'BOX OF VINTAGE ALPHABET BLOCKS',
  'PARTY BUNTING',
  'JAZZ HEARTS ADDRESS BOOK',
  'SET OF 4 SANTA PLACE SETTINGS'],
 'Quantity': [1, 24, 1, 2880, 2, 7, 1, 4, 10, 48],
 'UnitPrice': [1.65, 0.55, 1.65, 0.18, 1.25, 0.85, 11.95, 4.95, 0.19, 1.25],
 'Revenue': [1.65, 13.2, 1.65, 518.4, 2.5, 5.95, 11.95, 19.8, 1.9, 60.0]}

In [27]:
orders_df = pd.DataFrame(orders)
display(orders_df)


,Description,Quantity,UnitPrice,Revenue
0,LUNCH BAG APPLE DESIGN,1,1.65,1.65
1,SET OF 60 VINTAGE LEAF CAKE CASES,24,0.55,13.20
2,RIBBON REEL STRIPES DESIGN,1,1.65,1.65
3,WORLD WAR 2 GLIDERS ASSTD DESIGNS,2880,0.18,518.40
4,PLAYING CARDS JUBILEE UNION JACK,2,1.25,2.50
5,POPCORN HOLDER,7,0.85,5.95
6,BOX OF VINTAGE ALPHABET BLOCKS,1,11.95,11.95
7,PARTY BUNTING,4,4.95,19.80
8,JAZZ HEARTS ADDRESS BOOK,10,0.19,1.90
9,SET OF 4 SANTA PLACE SETTINGS,48,1.25,60.00


### 11. Calculate the total quantity ordered and revenue generated from these orders.

In [38]:
display(orders_df['Quantity'].sum())
display(orders_df['Revenue'].sum())



np.int64(2978)

np.float64(637.0)

### 12. Obtain the prices of the most expensive and least expensive items ordered and print the difference.

In [44]:
display(orders_df['UnitPrice'].max())
display(orders_df['UnitPrice'].min())
print(orders_df['UnitPrice'].max() - orders_df['UnitPrice'].min())

np.float64(11.95)

np.float64(0.18)

11.77


## Let's load another dataset for more exercisesº

In [45]:
# Run this code:
admissions = pd.read_csv('../Admission_Predict.csv')

Let's evaluate the dataset by looking at the `head` function.

In [47]:
display(admissions)

,Serial No.,GRE Score,TOEFL Score,University Rating,SOP,LOR,CGPA,Research,Chance of Admit
0,1,337,118,4,4.5,4.5,9.65,1,0.92
1,2,316,104,3,3.0,3.5,8.00,1,0.72
2,3,322,110,3,3.5,2.5,8.67,1,0.80
3,4,314,103,2,2.0,3.0,8.21,0,0.65
4,5,330,115,5,4.5,3.0,9.34,1,0.90
...,...,...,...,...,...,...,...,...,...
380,381,324,110,3,3.5,3.5,9.04,1,0.82
381,382,325,107,3,3.0,3.5,9.11,1,0.84
382,383,330,116,4,5.0,4.5,9.45,1,0.91
383,384,312,103,3,3.5,4.0,8.78,0,0.67


### 1 - Before beginning to work with this dataset and evaluating graduate admissions data, we will verify that there is no missing data in the dataset. Do this in the cell below.

In [51]:
admissions.isnull()

,Serial No.,GRE Score,TOEFL Score,University Rating,SOP,LOR,CGPA,Research,Chance of Admit
0,False,False,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False
3,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...
380,False,False,False,False,False,False,False,False,False
381,False,False,False,False,False,False,False,False,False
382,False,False,False,False,False,False,False,False,False
383,False,False,False,False,False,False,False,False,False


###  2 -  Interestingly, there is a column that uniquely identifies the applicants. This column is the serial number column. Instead of having our own index, we should make this column our index. Do this in the cell below. Keep the column in the dataframe in addition to making it an index.

In [64]:
admissions.set_index('Serial No.', drop = False)

,Serial No.,GRE Score,TOEFL Score,University Rating,SOP,LOR,CGPA,Research,Chance of Admit
Serial No.,,,,,,,,,
1,1,337,118,4,4.5,4.5,9.65,1,0.92
2,2,316,104,3,3.0,3.5,8.00,1,0.72
3,3,322,110,3,3.5,2.5,8.67,1,0.80
4,4,314,103,2,2.0,3.0,8.21,0,0.65
5,5,330,115,5,4.5,3.0,9.34,1,0.90
...,...,...,...,...,...,...,...,...,...
381,381,324,110,3,3.5,3.5,9.04,1,0.82
382,382,325,107,3,3.0,3.5,9.11,1,0.84
383,383,330,116,4,5.0,4.5,9.45,1,0.91


Turns out that `GRE Score` and `CGPA` also uniquely identify the data. Show this in the cell below.

### 3 - In this part of the lab, we would like to test complex conditions on the entire data set at once. Let's start by finding the number of rows where the CGPA is greater than 9 and the student has performed an investigation.

In [67]:
admissions = admissions.set_index(['GRE Score', 'CGPA'], drop=False)

### 4 - Now return all the rows where the CGPA is greater than 9 and the SOP score is less than 3.5. Find the mean chance of admit for these applicants.